providers/okta/application.go

```diff
@@ -0,0 +1,252 @@
+	// Prefer Duo OIDC flow
+	for _, rem := range resp.Remediation.Value {
+		if rem.Type == "OIDC" {
+			return DuoFrameless{remediation: rem}, true
+		}
+	}
+
+	for _, rem := range resp.Remediation.Value {
+		// This is intentionally 'Name' as Type is missing for cases where this is true.
+		if rem.Name == "challenge-authenticator" && resp.CurrentAuthenticatorEnrollment.Value.Key == "duo" {
+			host := resp.CurrentAuthenticatorEnrollment.Value.ContextualData.Host
+			tok, err := ParseSignedToken(resp.CurrentAuthenticatorEnrollment.Value.ContextualData.SignedToken)
+			if err != nil {
+				break
+			}
+
+			return DuoIframe{Host: host, SignedToken: tok, CallbackURL: rem.Href, Method: rem.Method, StateHandle: resp.StateHandle}, true
+		}
+	}
+
+	return nil, false
+}
+
+// ApplicationSAMLSource handles the SP-initiated SAML flow for KeyConjurer.
+type ApplicationSAMLSource string
+
+func (source ApplicationSAMLSource) URL() string {
+	return string(source)
+}
+
+func (source ApplicationSAMLSource) Identify(ctx context.Context, client *http.Client, prevURL *url.URL, username, password string, stateToken StateToken) (identifyResponse IdentifyResponse, err error) {
+	identifyVals := map[string]any{
+		"credentials": map[string]string{
+			"passcode": password,
+		},
+		"identifier":  username,
+		"stateHandle": stateToken,
+	}
+	identifyBytes, _ := json.Marshal(identifyVals)
+	identifyURL := prevURL.ResolveReference(&url.URL{Path: "/idp/idx/identify"})
+	resp, err := client.Post(identifyURL.String(), "application/json", bytes.NewReader(identifyBytes))
+	if err != nil {
+		return
+	}
+
+	if resp.StatusCode == http.StatusForbidden {
+		err = errors.New("user does not have access to this application")
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&identifyResponse)
+	return
+}
+
+type IntrospectResponse struct {
+	Success *struct {
+		Name string
+		Href string
+	}
+}
+
+func (source ApplicationSAMLSource) Introspect(ctx context.Context, client *http.Client, prevURL *url.URL, stateToken StateToken) (introspectResponse IntrospectResponse, err error) {
+	// We need to run introspect to get our next link.
+	// Now use /idp/idx/identify to get the IDP link to follow.
+	introspectVals := map[string]any{"stateToken": stateToken}
+	introspectBytes, _ := json.Marshal(introspectVals)
+	introspectURL := prevURL.ResolveReference(&url.URL{Path: "/idp/idx/introspect"})
+	resp, err := client.Post(introspectURL.String(), "application/json", bytes.NewReader(introspectBytes))
+	if err != nil {
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&introspectResponse)
+	return
+}
+
+func extractStateToken(r io.Reader) (StateToken, error) {
+	// Extremely cursed
+	bodyBuf, err := io.ReadAll(r)
+	if err != nil {
+		return "", err
+	}
+
+	// Extremely cursed
+	matches := stateTokenExpr.FindSubmatch(bodyBuf)
+	idx := stateTokenExpr.SubexpIndex("Token")
+	if len(matches) < idx+1 {
+		return "", errors.New("no match found")
```

Comment:
If there is a chance that this error is displayed to the user or in logs, I'd prefer to have a bit more details.

Reply:
if the error is caused by Okta changing something, hence breaking the application, I agree, it needs to be clear that support is needed.

Reply:
The details are listed at the call site.

```go
	stateToken, err := extractStateToken(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("could not find state token: %w", err)
	}
```